[fix][broker]Failed to create partitions after the partitions were deleted because topic GC

poorbarcode opened the pull request:
### Motivation

Configurations:
- `allowAutoTopicCreation`: `false`
- `brokerDeleteInactiveTopicsEnabled`: `true`
- `brokerDeleteInactivePartitionedTopicMetadataEnabled`: `false`

Steps of the issue occurred
- Create a topic with `2` partitions.
- Publish and consume messages.
- Stop all clients.
- `partition-0` is deleted because of no backlog.
- Restart clients failed, and get a `TopicNotFoundExpection`.


### Modifications

- Since the partitioned topic metadata still exists, broker should allow to load up partitions.

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

### Matching PR in forked repository

PR in forked repository: x

codelipenghui commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:
```diff
@@ -3614,11 +3614,20 @@ private CompletableFuture<Boolean> isAllowAutoTopicCreationAsync(final TopicName
+                    if (partitionedTopicMetadata.get().deleted) {
+                        return false;
+                    }
+                    return partitionedTopicMetadata.get().partitions + 1 >= topicName.getPartitionIndex();
```
It should be 
```suggestion
                    return partitionedTopicMetadata.get().partitions - 1 >= topicName.getPartitionIndex();
```

or just a simpler one

```suggestion
                    return partitionedTopicMetadata.get().partitions >  topicName.getPartitionIndex();
```

codelipenghui commented:
/pulsarbot run-failure-checks

coderzc commented:
@poorbarcode Please fix code style issue.

poorbarcode commented:
@nodece 

I changed the test that you added: `testCreateProducerWhenSinglePartitionIsDeleted`, because I have fixed the behaviour of the topic auto-creation.

Please take a look

nodece commented on pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerServiceAutoTopicCreationTest.java:
```diff
@@ -615,4 +619,53 @@ public void testAutoPartitionedTopicNameWithClusterName() throws Exception {
+                    .isDeleteWhileInactive());
+        });
+        // Trigger topic GC.
+        Awaitility.await().atMost(5, TimeUnit.SECONDS).untilAsserted(() -> {
+            persistentTopic.checkGC();
+            assertFalse(pulsar.getPulsarResources().getTopicResources()
+                    .persistentTopicExists(TopicName.get(topicP0)).get());
+        });
```
```suggestion
        persistentTopic.checkGC();
        Awaitility.await().atMost(5, TimeUnit.SECONDS).untilAsserted(() -> {
            assertFalse(pulsar.getPulsarResources().getTopicResources()
                    .persistentTopicExists(TopicName.get(topicP0)).get());
        });
```

poorbarcode replied:
It is beneficial to call GC multiple times as it avoids test instability caused by skipping execution due to certain edge conditions in a single call
